### ISSUE TITLE
onnxruntime-gpu, cudaoptions, result is different

### ISSUE THREAD

**Issue by snow-tyan** [microsoft/onnxruntime#13061]
### Describe the issue

When I use onnxruntime for inference, I found that the results of each run are different (to about 4 decimal places) after cuda options are used. Is this the accuracy problem caused by the copy from CPU to GPU? I used a for loop to verify this conjecture. I found that the position of the for loop would be like this before the session, but NOT after the session was created... Of course, if you turn off cuda and use cpu inference, the above problems will not occur in any case.

### To reproduce

```c++
    Ort::Env env(ORT_LOGGING_LEVEL_WARNING, "meter_recon");
    Ort::SessionOptions session_options;
    session_options.SetIntraOpNumThreads(NUM_THREADS);
    session_options.SetGraphOptimizationLevel(GraphOptimizationLevel::ORT_ENABLE_EXTENDED);
#define USE_CUDA
#ifdef USE_CUDA
    AF_INFO("USE CUDA, DEVICE_ID={:d}", DEVICE_ID);
    // OrtCUDAProviderOptions cuda_options;
    // cuda_options.device_id = DEVICE_ID;
    // cuda_options.cudnn_conv_algo_search = OrtCudnnConvAlgoSearchExhaustive;
    // cuda_options.gpu_mem_limit = std::numeric_limits<size_t>::max();
    // cuda_options.arena_extend_stracudategy = 0;
    // cuda_options.do_copy_in_default_stream = true;
    // session_options.AppendExecutionProvider_CUDA(cuda_options);
    OrtSessionOptionsAppendExecutionProvider_CUDA(session_options, DEVICE_ID);
#endif
for(int jjjj = 0; jjjj < 10; ++jjjj){
    Ort::Session session(env, model_path.c_str(), session_options);
    auto height = session.GetInputTypeInfo(0).GetTensorTypeAndShapeInfo().GetShape()[2];
    auto width = session.GetInputTypeInfo(0).GetTensorTypeAndShapeInfo().GetShape()[3];

    // Load Image
    cv::Mat in_mat = cv::imread(input_image.string());
...
}
```

### Platform

Linux

### OS Version

Docker: Ubuntu20.04

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.11.1-gpu-linux

### ONNX Runtime API

C++

### Architecture

X64

### Execution Provider

Default CPU, CUDA

### Execution Provider Library Version

CUDA 11.5, cuDNN8.3

**Comment by pranavsharma** [comment]
Can you paste the results from the different runs?

**Comment by snow-tyan** [comment]
> 你能粘贴不同运行的结果吗？

I rewrote a demo in Python and found that the same problem only occurred in CUDAExecutionProvider, but not in CPUExecutionProvider and TensorrtExecutionProvider.

I'm sorry that model weights and test images cannot be uploaded due to work reasons. However, I have tested the same problem with other models.

```shell
### Using CUDAExecutionProvider (Executed multiple times with the different result)
[301.5984001159668, 181.84381484985352, 51.91532, 68.349724, 0.9654644]
[301.59935569763184, 181.84461975097656, 51.91408, 68.34961, 0.96547186]
[301.59963607788086, 181.84502029418945, 51.913155, 68.34878, 0.9654721]
[301.5995502471924, 181.84505081176758, 51.913326, 68.34872, 0.9654721]
[301.59913063049316, 181.84360122680664, 51.914288, 68.35012, 0.9654635]

### Using CPUExecutionProvider (Executed multiple times with the same result)
[301.59936714172363, 181.8446159362793, 51.91412, 68.34959, 0.9654717]

### Using TensorrtExecutionProvider (Executed multiple times with the same result)
[301.5961723327637, 181.84202194213867, 51.91575, 68.35517, 0.9654652]
```

The following is the python test code

```python
import onnxruntime as ort

import cv2
from PIL import Image
import numpy as np
from typing import List
import os
import argparse


def PreProcess(input_data):
    input_data = input_data.astype('float32')

    mean = np.array([0., 0., 0.])
    std = np.array([255., 255., 255.])

    norm_input_data = np.zeros(input_data.shape).astype('float32')

    for i in range(input_data.shape[0]):
        norm_input_data[i, :, :] = ((input_data[i, :, :]) - mean[i]) / std[i]

    norm_input_data = norm_input_data.reshape(1, 3, 640, 640).astype('float32')
    return norm_input_data


def Nms(bboxes: List[List[float]], nms_thresh: float) -> None:
    bboxes.sort(key=lambda x: x[4], reverse=True)
    i = 0
    while i < len(bboxes):  ###! for range无法应对list变小的情况
        box_i = [bboxes[i][0], bboxes[i][1], bboxes[i][2], bboxes[i][3]]
        box_i_area = box_i[2] * box_i[3]
        j = i + 1
        while j < len(bboxes):
            box_j = [bboxes[j][0], bboxes[j][1], bboxes[j][2], bboxes[j][3]]
            box_j_area = box_j[2] * box_j[3]
            xx1 = max(box_i[0], box_j[0])
            yy1 = max(box_i[1], box_j[1])
            xx2 = min(box_i[0] + box_i[2], box_j[0] + box_j[2])
            yy2 = min(box_i[1] + box_i[3], box_j[1] + box_j[3])
            intersection = max(0., xx2 - xx1) * max(0., yy2 - yy1)
            union_f = box_i_area + box_j_area - intersection
            iou = intersection / union_f
            if (iou >= nms_thresh):
                bboxes.pop(j)
            else:
                j += 1
        i += 1


def ArgsParser():
    parser = argparse.ArgumentParser("meter_recon onnxruntime inference")
    parser.add_argument(
        '-m', '--model',
        default='models/sim/meter_det.sim.onnx',
        # default='models/remove_initializer/meter_det.onnx',
        type=str,
        help="model path")
    parser.add_argument(
        '-i', '--image',
        default='test_imgs/MeterRecon/cb1/20211102154627459008378666406441.jpg',
        type=str,
        help="image path")

    args = parser.parse_args()
    return args


if __name__ == '__main__':
    args = ArgsParser()

    in_mat = cv2.imread(args.image)
    # image = Image.open(image_path).resize((640, 640))
    in_mat = cv2.cvtColor(in_mat, cv2.COLOR_BGR2RGB)
    in_mat = cv2.resize(in_mat, (640, 640))
    image_data = np.array(in_mat).transpose(2, 0, 1)
    input_data = PreProcess(image_data)

    print(f'ort.get_device()={ort.get_device()}')
    print(f'ort.get_available_providers()={ort.get_available_providers()}')
    session_options = ort.SessionOptions()
    session_options.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_EXTENDED
    cpu_providers = ['CPUExecutionProvider']
    cuda_providers = [
        ('CUDAExecutionProvider', {
            'device_id': 0
        }),
        'CPUExecutionProvider',
    ]
    tensorrt_providers = [
        ('TensorrtExecutionProvider', {
            'device_id': 0,
            'trt_engine_cache_enable': True,
            'trt_engine_cache_path': './trtcache'
        }),
        ('CUDAExecutionProvider', {
            'device_id': 0
        }),
        'CPUExecutionProvider',
    ]
    ### 'gpu_mem_limit': '18446744073709551615' size_t::max (2^64-1)
    '''
    'providers' can contain either names or names and options. 
    When any options are given in 'providers', 'provider_options' should not be used.
    The list of providers is ordered by precedence. 
    For example `['CUDAExecutionProvider', 'CPUExecutionProvider']` means execute a node using `CUDAExecutionProvider`
    if capable, otherwise execute using `CPUExecutionProvider`.
    '''
    session = ort.InferenceSession(args.model,
                                   sess_options=session_options,
                                   providers=tensorrt_providers)
                                #    providers=cuda_providers)
                                #    providers=cpu_providers)
    print(session.get_providers())
    print(session.get_provider_options())

    ### inference
    # ortvalue = ort.OrtValue.ortvalue_from_numpy(input_data, 'cuda', 0)
    ortvalue = ort.OrtValue.ortvalue_from_numpy(input_data)
    print(ortvalue.device_name())  ## cpu or cuda
    print(ortvalue.shape())  ## [1, 3, 640, 640]
    print(ortvalue.data_type())  ## tensor(float)
    print(ortvalue.is_tensor())  ## True
    raw_result = session.run(['output'], {'images': ortvalue})
    # raw_result = session.run(['output'], {'images': ortvalue})
    print(f'raw_result[0].shape={raw_result[0].shape}')

    classes = 1
    background = -1
    total_classes = classes + (0 if background == -1 else -1)
    confidence_thresh = [0.3] * total_classes
    nms_thresh = [0.5] * total_classes
    boxes_number = raw_result[0].shape[1]
    print(f'boxes_number={boxes_number}')
    prediction = raw_result[0][0]  ## (1, 25200, 6)
    results = []  # [[x, y, w, h], score]
    for box_index in range(boxes_number):
        for total_class_index in range(total_classes):
            if background != -1 and total_class_index == background:
                continue
            class_index = total_class_index - 1 if background != -1 and total_class_index > background else total_class_index
            if prediction[box_index][4] > confidence_thresh[class_index]:
                center_x = prediction[box_index][0]
                center_y = prediction[box_index][1]
                w = prediction[box_index][2]
                h = prediction[box_index][3]
                results.append([
                    center_x - w / 2, center_y - h / 2, w, h,
                    prediction[box_index][4] * prediction[box_index][5]
                ])
    print(
        "Before NMS----------------------------------------------------------")
    print(f'result.size={len(results)}')
    # print(results)
    # print(results[0])
    # results[0].sort(key=lambda x: x[])
    if len(results) != 0:
        Nms(results, nms_thresh[0])
    print(
        "After NMS----------------------------------------------------------")
    print(f'result.size={len(results)}')
    print(results[0])
    # PostProcess()
```

**Comment by tianleiwu** [comment]
No, it is not related to data copy. It is likely caused by implementation of some operators, which need aggregation (like softmax, LayerNorm, Gemm). Since blocks of threads might be executed in different orders, so aggregation result might be slightly different. Running PyTorch model might also have similar issue.

To find the operator that causes the issue, you can refer to this code: https://github.com/microsoft/onnxconverter-common/blob/814cdf494d987900d30b16971c0e8334aaca9ae6/onnxconverter_common/auto_mixed_precision.py#L128-L148 to inspect the output of each node in multiple runs. Then you can do some statistics to see the first node that cause significant difference.

**Comment by snow-tyan** [comment]
> No, it is not related to data copy. It is likely caused by implementation of some operators, which need aggregation (like softmax, LayerNorm, Gemm). Since blocks of threads might be executed in different orders, so aggregation result might be slightly different. Running PyTorch model might also have similar issue.
> 
> To find the operator that causes the issue, you can refer to this code: https://github.com/microsoft/onnxconverter-common/blob/814cdf494d987900d30b16971c0e8334aaca9ae6/onnxconverter_common/auto_mixed_precision.py#L128-L148 to inspect the output of each node in multiple runs. Then you can do some statistics to see the first node that cause significant difference.

Thanks for reply. Does this problem occur even with a single thread?

**Comment by tianleiwu** [comment]
Cuda still use multiple threads even though you have only one cpu thread. 

This problem does not occur to every model so I suggest to investigate a little just in case it is a bug.

Another way is to add --cmake_extra_defines onnxruntime_DEBUG_NODE_INPUTS_OUTPUTS=1 in build command line to build a package from source code. Set environment variable  ORT_DEBUG_NODE_IO_DUMP_OUTPUT_DATA to be 1 before running your application. In this way, you can see of output of each node. Compare the stdout of two runs can find out the root cause.

**Comment by snow-tyan** [comment]
> Cuda still uses multiple threads even though you have only one cpu thread.
> 
> This problem does not occur to every model so I suggest to investigate a little just in case it is a bug.
> 
> Another way is to add --cmake_extra_defines onnxruntime_DEBUG_NODE_INPUTS_OUTPUTS=1 in build command line to build a package from source code. Set environment variable ORT_DEBUG_NODE_IO_DUMP_OUTPUT_DATA to be 1 before running your application. In this way, you can see of output of each node. Compare the stdout of two runs can find out the root cause.

Thanks again for your reply. I will recompile the ort to try the second method

**Comment by snow-tyan** [comment]
> Cuda still uses multiple threads even though you have only one cpu thread.
> 
> This problem does not occur to every model so I suggest to investigate a little just in case it is a bug.
> 
> Another way is to add --cmake_extra_defines onnxruntime_DEBUG_NODE_INPUTS_OUTPUTS=1 in build command line to build a package from source code. Set environment variable ORT_DEBUG_NODE_IO_DUMP_OUTPUT_DATA to be 1 before running your application. In this way, you can see of output of each node. Compare the stdout of two runs can find out the root cause.

I add --cmake_extra_defines onnxruntime_DEBUG_NODE_INPUTS_OUTPUTS=1 in build command.
And I use `ORT_DEBUG_NODE_IO_DUMP_OUTPUT_DATA=1 ./a.out` to see the output of each node successfully.
But when I use `ORT_DEBUG_NODE_IO_DUMP_OUTPUT_DATA=0 ./a.out`, it also prints the name of each node. How can I close this print information

**Comment by snow-tyan** [comment]
@tianleiwu I found that OrtCUDAProviderOptionsV2 was OK. And No previous problems. Thanks, I know the reason for the new version of optionsv2

**Comment by tianleiwu** [comment]
@snow-tyan, it is good to know that OrtCUDAProviderOptionsV2 is OK. Maybe it is caused by convolution search settings for cuDNN.
